src/modules/0.0.14/aws_s3/index.ts

```diff
@@ -1,63 +1,130 @@
+            bucket.createdAt = e.createdAt;
             await createBucket(client.s3Client, e.name);
+
+            out.push(bucket);
           }
+          return out;
         },
         read: async (ctx: Context, id?: string) => {
           const client = await ctx.getAwsClient() as AWS;
+
           const allBuckets = await getBuckets(client.s3Client);
-          return allBuckets
-            .filter(b => !id || b.Name === id)
-            .filter(b => !!b.Name)
-            .map(b => {
-              const bucket = new Bucket();
-              bucket.name = b.Name ?? ''; // The filter above is guarding this but TS is confused
-              bucket.createdAt = b.CreationDate;
-              return bucket;
-            });
+          const out : Bucket[] = [];
+          const buckets : BucketAWS[] = allBuckets
+              .filter(b => !id || b.Name === id)
+              .filter(b => !!b.Name);
+
+          for (const index in buckets ) {
+            if (buckets.hasOwnProperty(index)) {
+              const bucket = buckets[index];
+
+              // rerieve bucket policy
+              const input: GetBucketPolicyCommandInput = {
+                Bucket: bucket.Name,
+              };
+              const bucketPolicy = await getBucketPolicy(client.s3Client, input);
+
+              const b: Bucket = new Bucket();
+              b.name=bucket.Name!;
+              b.createdAt=bucket.CreationDate;
+
+              if (bucketPolicy) {
+                b.policyDocument=JSON.parse(bucketPolicy.Policy!);
+              } else {
+                b.policyDocument=undefined;
+              }
+              out.push(b);
+            }
+          }
+          return out;
+        },
+        updateOrReplace: (a: Bucket, b: Bucket) => {
+          if (!Object.is(a.policyDocument, b.policyDocument)) return 'update';
+          else return 'replace';
         },
         // TODO: With the model this simple it is actually impossible to really update this thing
         // as the user can't change the create timestamp themselves, so just restore the record from
         // the cloud cache and force it into the DB cache
         update: async (es: Bucket[], ctx: Context) => {
+          const client = await ctx.getAwsClient() as AWS;
           const out = [];
           for (const e of es) {
             const cloudRecord = ctx?.memo?.cloud?.Bucket?.[e.name ?? ''];
-            await AwsS3Module.mappers.bucket.db.update(cloudRecord, ctx);
-            ctx.memo.db.Bucket[cloudRecord.name] = cloudRecord;
-            out.push(cloudRecord);
+            const isUpdate = Object.is(AwsS3Module.mappers.bucket.cloud.updateOrReplace(cloudRecord, e), 'update');
+            if (isUpdate) {
+              const input : PutBucketPolicyCommandInput = {
+                Bucket: e.name,
+                Policy: JSON.stringify(e.policyDocument ?? {}),
+              };
+              await updateBucketPolicy(client.s3Client, input);
```

Comment:
Immediately after this line you need to insert these lines from the `cloud.read` path:

```suggestion
              await updateBucketPolicy(client.s3Client, input);
              const bucketPolicy = await getBucketPolicy(client.s3Client, { Bucket: e.name, });
              if (bucketPolicy) {
                e.policyDocument=JSON.parse(bucketPolicy.Policy!);
              } else {
                e.policyDocument=undefined;
              }
```